wolfHSM/src/chapter04.md

```diff
@@ -0,0 +1,274 @@
+are abstracted from the logical operations by associating callback functions
+with untyped context structures, referenced as a `void*`.
+
+### Example component initialization
+
+The prototypical compile-time static instance configuration and initialization
+sequence of a wolfHSM component is:
+
+```c
+#include "wolfhsm/component.h"        /* wolfHSM abstract API reference for a component */
+#include "port/vendor/mycomponent.h"  /* Platform specific definitions of configuration
+                                       * and context structures, as well as declarations of
+                                       * callback functions */ 
+
+/* Provide the lookup table for function callbacks for mycomponent. Note the type
+is the abstract type provided in wolfhsm/component.h */
+whComponentCb my_cb[1] = {MY_COMPONENT_CB};
+
+/* Fixed configuration data.  Note that pertinent data is copied out of the structure 
+ * during init() */
+const myComponentConfig my_config = {
+    .my_number = 3,
+    .my_string = "This is a string",
+}
+
+/* Static allocation of the dynamic state of the myComponent. */
+myComponentContext my_context[1] = {0};
+
+/* Initialization of the component using platform-specific callbacks */
+const whComponentConfig comp_config[1] = {
+        .cb = my_cb,
+        .context = my_context,
+        .config = my_config
+    };
+whComponentContext comp_context[1] = {0};
+int rc = wh_Component_Init(comp_context, comp_config);
+
+rc = wh_Component_DoSomething(comp_context, 1, 2, 3);
+rc = wh_Component_CleanUp(comp_context);
+```
+
+## Communications
+
+The communication layer of wolfHSM is designed to provide reliable, bidirectional, and packet-based communication between clients and servers. This layer abstracts the underlying transport mechanisms, allowing for flexibility and modularity. A key aspect of wolfHSM communication is split request and response functions for both client and server, enabling synchronous polling of message reception or asynchronous handling based on interrupt/event support.
+
+### Key Components
+
+- Client/Server APIs: Main interface for communicating between client and server. These are the APIs that are directly used by user applications.
+- Comms layer: Defines the format and structure of messages exchanged between clients and servers, and provides an abstract interface to the underlying transport layer implementation, exposing a consistent interface for sending and receiving messages.
+- Transport Layer: Concrete implementations of the underlying transport. Defines how data is actually transported between client and server.
+
+### Client/Server APIs:
+
+High level client and server APIs (defined in [wh_client.h](wolfhsm/wh_client.h) and [wh_server.h](wolfhsm/wh_server.h)) are the primary interface for communication. These functions abstract the low level communications details from the caller, providing a simple split transaction interface for logical operations.
```

Comment:
source file links don't work - these probably shoudn't be links, lets just make them code

```suggestion
High-level client and server APIs (defined in `wolfhsm/wh_client.h` and `wolfhsm/wh_server.h` are the primary interface for communication. These functions abstract the low level communications details from the caller, providing a simple split transaction interface for logical operations.
```